### PR TITLE
[WB-5669] Implement label proposal

### PR DESCRIPTION
**Pull request by raubitsj**
https://wandb.atlassian.net/browse/WB-5669

Description
-----------

Add ability to label scripts and repositories (and more eventually) with label tags in the source and embedded into frameworks.

This can be used to better understand how customers get activated: what was their initial experience? etc.

The format chosen was inspired by BibTex format.  The closest modern way of organizing code metadata is:
https://codemeta.github.io/, but there wasnt anything like an in-code annotation format that was terse enough.

How to use
-----------

Label tags can be added in multiple ways:
1. Add to the source script: (only checks the first docstring in the main module)
```python
"""
This is my doc string for my script

@wandbcode{myscript123, v=2}
"""
import wandb
```
2. Add to the notebook: (only checks the first cell)
3. Specify with `run._label()`

Analytics
---------

Label information is stored in the telemetry protocol buffer:
https://paper.dropbox.com/doc/CLI-Telemetry-Project--BEkdeh1pReI2bWOblq97J_xjAg-lPOOPwUGIJoMenU0DXJz4

Caveats
--------

This will only work for client library 0.10.32 and later, earlier client versions will not break, but they will not add label information.

Alternatives
------------

Store data directly in config:
PRO:  will work with all client library versions
CON: pollutes the wandb API, makes examples look more complicated.

Background
------------
Early tracking/label proposal:
https://paper.dropbox.com/doc/Tracking-examplesreposframeworks--BLgVczJOXu~9rXkBG6ijVFS7Ag-wSxbYSY5m3ZqmO5HjdIhP (outdated)

Future work
------------
- Encode common repos (this will allow for multiple repo labels in a run)
- Add usage info for improper (unhandled labels) - right now they are ignored
- Additional label parameters (checkpoints)

Testing
-------

How was this PR tested?

unittests

### CODE REVIEW

File: wandb/sdk_py27/wandb_run.py [hunk 1]
```diff
@@ -769,6 +771,67 @@ def entity(self):
...
+    # TODO: annotate jupyter Notebook class
+    def _label_probe_notebook(self, notebook):
+        logger.info("probe notebook")
+        lines = None
+        try:
+            data = notebook.probe_ipynb()
+            cell0 = data.get("cells", [])[0]
+            lines = cell0.get("source")
+        except Exception as e:
+            print("ERR", e)
```
Comment (dannygoldstein) [inline wandb/sdk_py27/wandb_run.py:829]:
```suggestion
            termerror(e)
```

File: wandb/sdk/wandb_run.py [hunk 2]
```diff
@@ -769,6 +771,67 @@ def entity(self) -> str:
...
+    # TODO: annotate jupyter Notebook class
+    def _label_probe_notebook(self, notebook: Any) -> None:
+        logger.info("probe notebook")
+        lines = None
+        try:
+            data = notebook.probe_ipynb()
+            cell0 = data.get("cells", [])[0]
+            lines = cell0.get("source")
+        except Exception as e:
+            print("ERR", e)
```
Comment (dannygoldstein) [inline wandb/sdk/wandb_run.py:829]: see below

**Review by dannygoldstein — APPROVED** [review]
Looks great! Looking forward to this. How will the id be used on the backend? If a single traceback touches frames with multiple `@wandbcode` labels (e.g., label on the user main script, and label on the pytorch framework or similar), will those labels be combined in sentry?

File: wandb/sdk_py27/wandb_run.py (hunk 1 above)
Comment (raubitsj) [reply wandb/sdk_py27/wandb_run.py:829]: thanks. this is supposed to be silent, so just the logger is good enough. fixd

**Comment by raubitsj** [comment]
> Looks great! Looking forward to this. How will the id be used on the backend? If a single traceback touches frames with multiple `@wandbcode` labels (e.g., label on the user main script, and label on the pytorch framework or similar), will those labels be combined in sentry?

Good question.  There will be a new field added to the Label proto that has all the partner repos, the string will catch the new ones.  I will put this in a follow on PR, not sure if it will make this release or next.
There will only be one "code" string, since it is the main program only.

**Pull request merged**